modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManagerImpl.java

```diff
@@ -17,53 +17,282 @@
+
+                        String placeholderValue = keyTail.substring(0, keyTail.indexOf('.'));
+
+                        UUID tblId = UUID.fromString(placeholderValue);
+
+                        try {
+                            String name = new String(metaStorageMgr.get(
+                                new Key(INTERNAL_PREFIX + tblId.toString())).get()
+                                .value(), StandardCharsets.UTF_8);
+                            int partitions = configurationMgr.configurationRegistry().getConfiguration(TablesConfiguration.KEY)
+                                .tables().get(name).partitions().value();
+
+                            List<List<ClusterNode>> assignment = (List<List<ClusterNode>>)IgniteUtils.fromBytes(
+                                evt.newEntry().value());
+
+                            HashMap<Integer, RaftGroupService> partitionMap = new HashMap<>(partitions);
+
+                            for (int p = 0; p < partitions; p++) {
+                                partitionMap.put(p, raftMgr.startRaftGroup(
+                                    name + "_part_" + p,
+                                    assignment.get(p),
+                                    new PartitionCommandListener()
+                                ));
+                            }
+
+                            tables.put(name, new TableImpl(
+                                new InternalTableImpl(
+                                    tblId,
+                                    partitionMap,
+                                    partitions
+                                ),
+                                new TableSchemaView() {
+                                    @Override public SchemaDescriptor schema() {
+                                        return schemaManager.schema(tblId);
+                                    }
+
+                                    @Override public SchemaDescriptor schema(int ver) {
+                                        return schemaManager.schema(tblId, ver);
+                                    }
+                                }));
+                        }
+                        catch (InterruptedException | ExecutionException e) {
+                            LOG.error("Failed to start table [key={}]",
+                                evt.newEntry().key(), e);
+                        }
+                    }
+                }
+
+                return false;
+            }
+
+            @Override public void onError(@NotNull Throwable e) {
+                LOG.error("Metastorage listener issue", e);
+            }
+        });
+    }
+
+    /**
+     * Checks whether the local node hosts Metastorage.
+     *
+     * @param localNodeName Local node uniq name.
+     * @param metastorageMembers Metastorage members names.
+     * @return True if the node has Metastorage, false otherwise.
+     */
+    private boolean hasMetastorageLocally(String localNodeName, String[] metastorageMembers) {
```

Comment:
Seems that we have this logic both in AffinityManager and TableManager. Let's extract it somewhere For example to MetaStorageManager.